src/kernel/network.cpp

```diff
@@ -55,211 +147,183 @@ CryptoKernel::Network::Network(CryptoKernel::Log* log,
-                                    Json::Value newSeed;
-                                    newSeed["lastseen"] = 0;
-                                    newSeed["height"] = 1;
-                                    newSeed["score"] = 0;
-                                    peers->put(dbTx.get(), addr.toString(), newSeed);
-                                }
-                            } else {
-                                changeScore(it->first, 10);
-                                throw Peer::NetworkError();
-                            }
-                        }
-                    } catch(const Json::Exception& e) {
-                        changeScore(it->first, 50);
-                        throw Peer::NetworkError();
-                    }
-
-                    const std::time_t result = std::time(nullptr);
-                    it->second->info["lastseen"] = static_cast<uint64_t>(result);
-                } catch(const Peer::NetworkError& e) {
-                    log->printf(LOG_LEVEL_WARN,
-                                "Network(): Failed to contact " + it->first + ", disconnecting it");
-                    removals.insert(it->first);
-                }
-            }
-
-            for(const auto& peer : removals) {
-                const auto it = connected.find(peer);
-                peers->put(dbTx.get(), peer, it->second->info);
-                if(it != connected.end()) {
-                    connected.erase(it);
-                }
-            }
-
-            for(const auto& peer : peerInfos) {
-                peers->put(dbTx.get(), peer.first, peer.second);
-            }
+void CryptoKernel::Network::makeOutgoingConnectionsWrapper() {
+	while(running) {
+		makeOutgoingConnections();
+		std::this_thread::sleep_for(std::chrono::milliseconds(100));
+	}
+}
 
-            dbTx->commit();
+void CryptoKernel::Network::infoOutgoingConnectionsWrapper() {
+	while(running) {
+		bool wait = false;
 
-            std::lock_guard<std::mutex> cLock(connectedStatsMutex);
-            connectedStats.clear();
+		infoOutgoingConnections();
+		std::this_thread::sleep_for(std::chrono::milliseconds(100));
+	}
+}
 
-            for(const auto& peer : connected) {
-                peerStats stats = peer.second->peer->getPeerStats();
-                stats.version = peer.second->info["version"].asString();
-                stats.blockHeight = peer.second->info["height"].asUInt64();
-                connectedStats.insert(std::make_pair(peer.first, stats));
-            }
-        }
+void CryptoKernel::Network::makeOutgoingConnections() {
+	std::map<std::string, Json::Value> peersToTry;
+	{
+		CryptoKernel::Storage::Table::Iterator* it = new CryptoKernel::Storage::Table::Iterator(
+				peers.get(), networkdb.get());
+
+		for(it->SeekToFirst(); it->Valid(); it->Next()) {
+			if(connected.size() >= 8) { // honestly, this is enough
+				std::this_thread::sleep_for(std::chrono::seconds(20)); // so stop looking for a while
```

Comment:
`makeOutgoingConnections` sleeps for 20 seconds after 8 connected peers while the `Storage::Table::Iterator` still exists and thus the db is locked. That means `infoOutgoingConnections` is blocking during this process as it can't begin a db transaction until the iterator is deleted.

Should break instead and the sleep be moved to the outer loop.